cmd/kube-controller-manager/app/certificates.go

```diff
@@ -89,22 +91,38 @@ func startCSRSigningController(ctx ControllerContext) (http.Handler, bool, error
 	}
 
 	c := ctx.ClientBuilder.ClientOrDie("certificate-controller")
+	csrInformer := ctx.InformerFactory.Certificates().V1beta1().CertificateSigningRequests()
+	certTTL := ctx.ComponentConfig.CSRSigningController.ClusterSigningDuration.Duration
+	caFile, caKeyFile := getKubeletServingSignerFiles(ctx.ComponentConfig.CSRSigningController)
 
-	signer, err := signer.NewCSRSigningController(
-		c,
-		ctx.InformerFactory.Certificates().V1beta1().CertificateSigningRequests(),
-		ctx.ComponentConfig.CSRSigningController.ClusterSigningCertFile,
-		ctx.ComponentConfig.CSRSigningController.ClusterSigningKeyFile,
-		ctx.ComponentConfig.CSRSigningController.ClusterSigningDuration.Duration,
-	)
-	if err != nil {
-		return nil, false, fmt.Errorf("failed to start certificate controller: %v", err)
+	if signer, err := signer.NewKubeletServingCSRSigningController(c, csrInformer, caFile, caKeyFile, certTTL); err != nil {
+		return nil, false, fmt.Errorf("failed to start kubernetes.io/kubelet-serving certificate controller: %v", err)
+	} else {
+		go signer.Run(1, ctx.Stop)
+	}
+	if signer, err := signer.NewKubeletClientCSRSigningController(c, csrInformer, caFile, caKeyFile, certTTL); err != nil {
+		return nil, false, fmt.Errorf("failed to start kubernetes.io/kube-apiserver-client-kubelet certificate controller: %v", err)
+	} else {
+		go signer.Run(1, ctx.Stop)
+	}
+	if signer, err := signer.NewKubeAPIServerClientCSRSigningController(c, csrInformer, caFile, caKeyFile, certTTL); err != nil {
+		return nil, false, fmt.Errorf("failed to start kubernetes.io/kube-apiserver-client certificate controller: %v", err)
+	} else {
+		go signer.Run(1, ctx.Stop)
+	}
+	if signer, err := signer.NewLegacyUnknownCSRSigningController(c, csrInformer, caFile, caKeyFile, certTTL); err != nil {
+		return nil, false, fmt.Errorf("failed to start kubernetes.io/legacy-unknown certificate controller: %v", err)
+	} else {
+		go signer.Run(1, ctx.Stop)
 	}
-	go signer.Run(1, ctx.Stop)
 
 	return nil, true, nil
 }
 
+func getKubeletServingSignerFiles(config csrsigningconfig.CSRSigningControllerConfiguration) (string, string) {
+	return config.ClusterSigningCertFile, config.ClusterSigningKeyFile
+}
```

Comment:
What's the plan for this?

Reply:
Presumably we would wire up different CLI flags to each signer since each can have a distinct cert and key file.

Reply:
> Presumably we would wire up different CLI flags to each signer since each can have a distinct cert and key file.

yes, with a priority so the old flag works.